runners/flink/src/main/java/org/apache/beam/runners/flink/execution/EnvironmentSession.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.runners.flink.execution;
+
+import org.apache.beam.model.pipeline.v1.RunnerApi.Environment;
+import org.apache.beam.runners.fnexecution.artifact.ArtifactSource;
+import org.apache.beam.runners.fnexecution.control.SdkHarnessClient;
+
+/**
+ * A handle to an {@link Environment} managed by a {@link SdkHarnessManager}.
+ * Closing the session indicates to the session's {@link SdkHarnessManager}
+ * that it can no longer use session resources such as its
+ * {@link ArtifactSource}.
+ */
+public interface EnvironmentSession extends AutoCloseable {
```

Comment:
It looks like this may duplicate the functionality of `RemoteEnvironment` in apache/beam#4589. You might want to consider patching that in, waiting for it to merge, or duplicating that interface here to make switching easier.

Reply:
resolved offline: Whereas RemoteEnvironment encapsulates the lifetime of an SDK container, this encapsulates the lifetime of a FlinkOperator's activity.  The key distinction is that when you close a RemoteEnvironment, the SDK container shuts down.  When you close an EnvironmentSession, the SDK container stays up, but the ArtifactServer is notified that this session's ArtifactSource cannot serve further requests.